File: tree/144/144.py
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        ans = []
        self.traverse(root, ans)
        return ans

    def traverse(self, root, ans):
        if root is None:
            return
        ans.append(root.val)
        self.traverse(root.left, ans)
        self.traverse(root.right, ans)

# solution 2


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        ans = []
        stack = []
        if root:
            stack.append(root)
        while stack:
            cur = stack.pop()
            ans.append(cur.val)
            if cur.right:
                stack.append(cur.right)
            if cur.left:
                stack.append(cur.left)
        return ans

File: tree/144/144.js
// solution 1
var preorderTraversal = function(root) {
  let ans = [];
  traverse(root, ans);
  return ans;
};

const traverse = (root, ans) => {
  if (root === null) return;
  ans.push(root.val);
  traverse(root.left, ans);
  traverse(root.right, ans);
};

// solution 2
var preorderTraversal = function(root) {
  let ans = [];
  let stack = [];
  if (root) stack.push(root);
  while (stack.length !== 0) {
    const curr = stack.pop();
    ans.push(curr.val);
    if (curr.right) stack.push(curr.right);
    if (curr.left) stack.push(curr.left);
  }
  return ans;
};
